napari/utils/events/dataclass.py

```diff
@@ -0,0 +1,421 @@
+        out[name] = TypeGetSet(*d)
+    return out
+
+
+def convert_fields_to_properties(cls: Type[C]) -> Type[C]:
+    """Convert all fields in a dataclass instance to property descriptors.
+
+    Note: this modifies class Type[C] (the class that was decorated with
+    ``@dataclass``) *after* instantiation of the class.  In other words, for a
+    given field `f` on class `C`, `C.f` will *not* be a property descriptor
+    until *after* C has been instantiated: `c = C()`.  (And reminder: property
+    descriptors are class attributes).
+
+    The reason for this is that dataclasses can have "default factory"
+    functions that create default values for fields only during instantiation,
+    and we don't want to have to recreate that logic here, (but we do need to
+    know what the value of the field is).
+
+    Parameters
+    ----------
+    obj : T
+        An instance of class ``Type[C]`` that has been deorated as a dataclass.
+    """
+    from numpydoc.docscrape import ClassDoc
+
+    # grab docstring to populate properties docs
+    cls_doc = ClassDoc(cls)
+    params = {p.name: p for p in cls_doc["Parameters"]}
+
+    coerce_funcs = parse_annotated_types(cls)
+    # loop through annotated members of the glass
+    for name, type_ in list(cls.__dict__.get('__annotations__', {}).items()):
+        # ClassVar and InitVar types are exempt from dataclasses and properties
+        # https://docs.python.org/3/library/dataclasses.html#class-variables
+        if _dc._is_classvar(type_, typing) or _dc._is_initvar(type_, _dc):
+            continue
+        private_name = f"_{name}"
+        # store the original value for the property
+        default = getattr(cls, name, _dc.MISSING)
+
+        # add the private_name as a ClassVar annotation on the original class
+        # (annotations of type ClassVar are ignored by the dataclass)
+        cls.__annotations__[private_name] = ClassVar[type_]
+
+        fget = prop_getter(private_name, coerce_funcs.get(name).fget)
+        fset = prop_setter(private_name, default, coerce_funcs.get(name).fset)
+        # bring the docstring from the class to the property
+        doc = None
+        if name in params:
+            param = params[name]
+            doc = "\n".join(param.desc)
+            # TODO: could compare param.type to field.type here for consistency
+            # alternatively, we may just want to use pydantic for type
+            # validation.
+
+        # create the actual property descriptor
+        prop = property(fget=fget, fset=fset, fdel=None, doc=doc)
+        setattr(cls, name, prop)
+    return cls
+
+
+class Property:
+    """Declare a dataclass field as a property with getter/setter functions"""
+
+    def __new__(cls, *args, **kwargs):
+        raise TypeError("Type Property cannot be instantiated")
+
+    def __init_subclass__(cls, *args, **kwargs):
+        raise TypeError(f"Cannot subclass {cls.__module__}.Property")
+
+    import typing_extensions
```

Comment:
is there a reason for importing this despite there being `_te`? if so, could you comment why please?